packages/accounts/src/msca/builder.ts

```diff
@@ -0,0 +1,151 @@
+) => Promise<Hex>;
+
+// TODO: this can be moved out into its own file
+export const StandardExecutor: Executor = () => ({
+  async encodeExecute(
+    target: Address,
+    value: bigint,
+    data: Hex
+  ): Promise<`0x${string}`> {
+    return encodeFunctionData({
+      abi: IStandardExecutorAbi,
+      functionName: "execute",
+      args: [target, value, data],
+    });
+  },
+
+  async encodeBatchExecute(
+    txs: BatchUserOperationCallData
+  ): Promise<`0x${string}`> {
+    return encodeFunctionData({
+      abi: IStandardExecutorAbi,
+      functionName: "executeBatch",
+      args: [
+        txs.map((tx) => ({
+          target: tx.target,
+          data: tx.data,
+          value: tx.value ?? 0n,
+        })),
+      ],
+    });
+  },
+});
+
+const zCompleteBuilder = z.object({
+  executor: z.custom<Executor>(),
+  signer: z.custom<SignerMethods>(),
+  factory: z.custom<Factory>(),
+});
+
+export class MSCABuilder {
+  executor?: Executor;
+  signer?: SignerMethods;
+  factory?: Factory;
+
+  withExecutor(executor: Executor): this & { executor: Executor } {
+    return Object.assign(this, { executor });
+  }
+
+  withSigner(methods: SignerMethods): this & { signer: SignerMethods } {
+    return Object.assign(this, { signer: methods });
+  }
+
+  withFactory(initCode: Factory): this & { factory: Factory } {
+    return Object.assign(this, { factory: initCode });
+  }
+
+  build<TTransport extends SupportedTransports = Transport>(
+    params: BaseSmartAccountParams
+  ): MSCA {
+    const builder = this;
+    const { signer, executor, factory } = zCompleteBuilder.parse(builder);
+
+    return new (class extends BaseSmartContractAccount<TTransport> {
+      getDummySignature(): `0x${string}` {
+        return signer(this, this.rpcProvider).getDummySignature();
+      }
+
+      encodeExecute(
+        target: string,
+        value: bigint,
+        data: string
+      ): Promise<`0x${string}`> {
+        return executor(this, this.rpcProvider).encodeExecute(
+          target,
+          value,
+          data
+        );
+      }
+
+      encodeBatchExecute(
+        txs: BatchUserOperationCallData
+      ): Promise<`0x${string}`> {
+        return executor(this, this.rpcProvider).encodeBatchExecute(txs);
+      }
+
+      signMessage(msg: string | Uint8Array): Promise<`0x${string}`> {
+        return signer(this, this.rpcProvider).signMessage(msg);
+      }
+
+      signTypedData(params: SignTypedDataParams): Promise<`0x${string}`> {
+        return signer(this, this.rpcProvider).signTypedData(params);
+      }
+
+      signUserOperationHash(uoHash: `0x${string}`): Promise<`0x${string}`> {
+        return signer(this, this.rpcProvider).signUserOperationHash(uoHash);
+      }
+
+      protected getAccountInitCode(): Promise<`0x${string}`> {
+        return factory(this, this.rpcProvider);
+      }
+
+      extendWithPluginMethods = <D>(plugin: Plugin<D>): this & D => {
```

Comment:
decorators should probably be a method that takes in the account as param (optionally) and returns the methods

Reply:
breaking down a plugin's methods further...

There are probably methods that allow you to:
1. helpers on the ABI that allow you to encode data for calls to the contract (current decorator methods)
2. decorators that are meant to be added to an `ISmartAccountProvider` to add ergo utils like sending UOs that aren't encoded to the execute methods
3. decorators meant to be added to a `SmartContractAccount` that make it easier to make eth_calls

1 and 3 are very similar, so I think we probably want to update the plugin type to be something liek 

```ts
export interface Plugin<AD, PD> {
  meta: { name: string; version: string };
  accountDecorators: (a: SmartContractAccount) => AD;
  providerDecorators: (p: ISmartAccountProvider) => PD;
}
```

For example, even though I have the decorators exported, I couldn't actually make use of them when defining the `sign*` methods for MSCA in this PR

Reply:
Addressed this in this PR